torch/csrc/jit/tensorexpr/reduction.h

```diff
@@ -120,6 +120,42 @@ class TORCH_API Reducer {
   ReduceInteraction interaction_;
 };
 
+class ReduceXOp : public ExprNode<ReduceXOp> {
```

Comment:
That's a somewhat dangerous introduction. If we decide to go with it, we'd need to define what happens in cases when an expression contains several reduce sub expressions.
Imagine, for instance, that we'd like to represent the following expression:
```
a[i] = sum(b[i,j,k], reduce_axis=[j,k]) + sum(c[i,l,m,p,q], reduce_axis=[l,m,p,q])
```
Now to lower this to statements, we'd need to generate 4 loops for the second reduction and 2 loops for the first one. Before this change, we implicitly enforced that by the fact that Reduce could only produce a statement - if one would like to use results of several reductions, they would need to create a separate "computation" which would mean a separate nest of loops. Now we would need some machinery to make sure it happens automatically and correctly.

I'd love to hear what you think about such cases - how would we handle them?

Reply:
This is a great point.  We actually don't want to support reductions below the top level; a nice thing about the current interface is that you can't express them!  But unfortunately it prevents us from expressing things like data-dependent reduction axes :-/.

The best thing I can think of is to add a dynamic check that reduction only occur at the top level.